domain/machine/state/state.go

```diff
@@ -36,64 +36,140 @@ func NewState(factory coredb.TxnRunnerFactory, logger logger.Logger) *State {
+	createNodeStmt, err := st.Prepare(createNodeQuery, createParams)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	createParams := sqlair.M{
-		"machine_uuid":  machineUUID,
-		"net_node_uuid": nodeUUID,
-		"name":          machineName,
-		"life_id":       life.Alive,
+	// Prepare query for associating parent machine.
+	var parentNameParam machineName
+	var associateParentStmt *sqlair.Statement
+	var associateParentParam machineParent
+	if args.parentName != "" {
+		parentNameParam = machineName{Name: args.parentName}
+		associateParentParam = machineParent{MachineUUID: args.machineUUID}
+		associateParentQuery := `
+INSERT INTO machine_parent (machine_uuid, parent_uuid)
+VALUES ($machineParent.machine_uuid, $machineParent.parent_uuid)
+`
+		associateParentStmt, err = st.Prepare(associateParentQuery, associateParentParam)
+		if err != nil {
+			return errors.Trace(err)
+		}
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		result := sqlair.M{}
-		err := tx.Query(ctx, queryStmt, machineNameParam).Get(&result)
-		if err != nil {
-			if !errors.Is(err, sqlair.ErrNoRows) {
-				return errors.Annotatef(err, "querying machine %q", machineName)
-			}
-		}
-		// For now, we just care if the minimal machine row already exists.
+		// Query for the machine uuid. If the machine already exists, return a
+		// MachineAlreadyExists error.
+		err := tx.Query(ctx, machineUUIDStmt, machineNameParam).Get(&machineUUIDout)
+		// No error means we found the machine with the given name.
 		if err == nil {
-			return nil
+			return errors.Annotatef(machineerrors.MachineAlreadyExists, "machine %q", mName)
+		} else if !errors.Is(err, sqlair.ErrNoRows) {
+			// Return error if the query failed for any reason other than not
+			// found.
+			return errors.Annotatef(err, "querying machine %q", mName)
 		}
 
+		// Run query to create net node row.
 		if err := tx.Query(ctx, createNodeStmt, createParams).Run(); err != nil {
-			return errors.Annotatef(err, "creating net node row for machine %q", machineName)
+			return errors.Annotatef(err, "creating net node row for machine %q", mName)
 		}
+
+		// Run query to create machine row.
 		if err := tx.Query(ctx, createMachineStmt, createParams).Run(); err != nil {
-			return errors.Annotatef(err, "creating machine row for machine %q", machineName)
+			return errors.Annotatef(err, "creating machine row for machine %q", mName)
 		}
+
+		// Associate a parent machine if parentName is provided.
+		if args.parentName != "" {
+			// Query for the parent uuid.
+			// Reusing the machineUUIDout variable for the parent.
+			err := tx.Query(ctx, machineUUIDStmt, parentNameParam).Get(&machineUUIDout)
+			if err != nil {
+				if errors.Is(err, sqlair.ErrNoRows) {
+					return errors.NotFoundf("parent machine %q for %q", args.parentName, mName)
+				}
+				return errors.Annotatef(err, "querying parent machine %q for machine %q", args.parentName, mName)
+			}
```

Comment:
**suggestion:** I don't like nested if. For me this is an error handling from mor specific to less specific.

so we can write:

```go
if errors.Is(err, sqlair.ErrNoRows) {
    return errors.NotFoundf("parent machine %q for %q", args.parentName, mName)
}
if err != nil {    
    return errors.Annotatef(err, "querying parent machine %q for machine %q", args.parentName, mName)
}
```